singularity-eos/eos/eos_stellar_collapse.hpp

```diff
@@ -819,23 +879,61 @@ inline void StellarCollapse::readBounds_(const hid_t &file_id, const std::string
  * https://forum.hdfgroup.org/t/is-this-a-bug-in-hdf5-1-8-6/2211
  */
 inline void StellarCollapse::readSCDset_(const hid_t &file_id, const std::string &name,
-                                         DataBox &db) {
+                                         const Grid_t &Ye_grid, const Grid_t &lT_grid,
+                                         const Grid_t &lRho_grid, DataBox &db) {
   herr_t exists = H5LTfind_dataset(file_id, name.c_str());
   if (!exists) {
     std::string msg = "Tried to read dataset " + name + " but it doesn't exist\n";
     EOS_ERROR(msg.c_str());
   }
-  db.resize(numYe_, numT_, numRho_);
+  db.resize(Ye_grid.nPoints(), lT_grid.nPoints(), lRho_grid.nPoints());
   herr_t status = H5LTread_dataset(file_id, name.c_str(), H5T_REAL, db.data());
   if (status != H5_SUCCESS) {
     std::string msg = "Tried to read dataset " + name + " but it failed exist\n";
-    EOS_ERROR(msg.c_str());
+    PORTABLE_ALWAYS_THROW_OR_ABORT(msg.c_str());
   }
 
   // bounds
-  db.setRange(2, YeMin_, YeMax_, numYe_);
-  db.setRange(1, lTMin_, lTMax_, numT_);
-  db.setRange(0, lRhoMin_, lRhoMax_, numRho_);
+  db.setRange(2, Ye_grid);
+  db.setRange(1, lT_grid);
+  db.setRange(0, lRho_grid);
+}
+
+// Reinterpolate tab form its original grid spacing to the one using
+// the native log gridding for stellar collapse (usually fast logs).
+// Scratch is used as a temporary storage buffer and is assumed to be
+// of size numYe x numT x numRho
+inline void StellarCollapse::reinterpolateTable_(DataBox &tab, DataBox &scratch,
+                                                 bool dependent_var_log) {
+  // Start by assuming tab is gridded for Ye, log10(T), log10(rho)
+  for (int iY = 0; iY < numYe_; ++iY) {
+    Real Ye = YeGrid_.x(iY);
+    for (int iT = 0; iT < numT_; ++iT) {
+      Real lT = lTGrid_.x(iT);
+      Real T = T_(lT);
+      Real log10T = std::log10(T);
+      for (int iRho = 0; iRho < numRho_; ++iRho) {
+        Real lR = lRGrid_.x(iRho);
+        Real rho = rho_(lR);
+        Real log10R = std::log10(rho);
+        Real val = tab.interpToReal(Ye, log10T, log10R);
+        if (dependent_var_log) {
+          // even if there is an offset, it's applied to the linear
+          // term before taking the logarithm, so don't need to
+          // include it here.
+          val = toLog_(std::pow(10., val), 0);
+        }
+        scratch(iY, iT, iRho) = val;
+      }
+    }
+  }
+  // loop through a second time and copy
+  for (int i = 0; i < numYe_ * numT_ * numRho_; ++i) {
+    tab(i) = scratch(i);
+  }
+  tab.setRange(2, YeGrid_);
+  tab.setRange(1, lTGrid_);
+  tab.setRange(0, lRGrid_);
```

Comment:
Does this function deserve a unit test? I'm assuming you already checked it for accuracy elsewhere

Reply:
You mean `setRange`? That's comprehensively tested in `spiner`.

Reply:
Oh---no I haven't carefully tested it. It probably does deserve its own unit test, though.

Reply:
Added a test.